File: LeetCode/2095. Delete the Middle Node of a Linked List.js
// Author : @AdarshGaur

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteMiddle = function(head) {
    if(head.next == null)
        return null
    
    let tortoise = head, hare = head.next, prev = null
    while(hare != null && hare.next != null)
        prev = tortoise, tortoise = tortoise.next, hare = hare.next.next
    
    let left = tortoise, right = tortoise.next.next
    if(hare == null)
        left = prev, right = tortoise.next
    
    left.next = null
    left.next = right
    
    return head
};

File: LeetCode/2095. Delete the Middle Node of a Linked List.py
# Author : @AdarshGaur

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteMiddle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        
        if head.next == None: return None
        
        tortoise = head
        hare = head.next
        prev = None
        
        while hare != None and hare.next != None:
            hare = hare.next.next
            prev = tortoise
            tortoise = tortoise.next
        
        left = tortoise
        right = tortoise.next.next
        
        if hare == None:
            left = prev
            right = tortoise.next
        
        left.next = right
        return head
        
